src/descriptor/policy.rs

```diff
@@ -787,3 +787,344 @@ impl ExtractPolicy for Descriptor<DescriptorPublicKey> {
+            .unwrap()
+            .unwrap();
+
+        assert!(
+            matches!(&policy.item, Multisig { keys, threshold } if threshold == &1
+            && &keys[0].fingerprint.unwrap() == &fingerprint0
+            && &keys[1].fingerprint.unwrap() == &fingerprint1)
+        );
+        assert!(
+            matches!(&policy.contribution, Satisfaction::PartialComplete { n, m, items, conditions } if n == &2
+             && m == &1
+             && items.len() == 2
+             && conditions.contains_key(&vec![0])
+             && conditions.contains_key(&vec![1])
+            )
+        );
+    }
+
+    // test ExtractPolicy trait with descriptors containing timelocks in a thresh()
+
+    #[test]
+    fn test_extract_policy_for_wsh_multi_timelock() {
+        let (prvkey0, _pubkey0, _fingerprint0) = setup_keys(TPRV0_STR);
+        let (_prvkey1, pubkey1, _fingerprint1) = setup_keys(TPRV1_STR);
+        let sequence = 50;
+        let desc = descriptor!(wsh (
+            thresh 2, (pk prvkey0), (+s pk pubkey1), (+s+d+v older sequence)
+        ))
+        .unwrap();
+
+        let (wallet_desc, keymap) = desc.to_wallet_descriptor(Network::Testnet).unwrap();
+        let signers_container = Arc::new(SignersContainer::from(keymap));
+        let policy = wallet_desc
+            .extract_policy(signers_container)
+            .unwrap()
+            .unwrap();
+
+        assert!(
+            matches!(&policy.item, Thresh { items, threshold } if items.len() == 3 && threshold == &2)
+        );
+
+        assert!(
+            matches!(&policy.contribution, Satisfaction::PartialComplete { n, m, items, conditions } if n == &3
+             && m == &2
+             && items.len() == 3
+             && conditions.get(&vec![0,1]).unwrap().iter().next().unwrap().csv.is_none()
+             && conditions.get(&vec![0,2]).unwrap().iter().next().unwrap().csv == Some(sequence)
+             && conditions.get(&vec![1,2]).unwrap().iter().next().unwrap().csv == Some(sequence)
+            )
+        );
+    }
+
+    // - mixed timelocks should fail
+
+    #[test]
+    fn test_extract_policy_for_wsh_mixed_timelocks() {
+        let (prvkey0, _pubkey0, _fingerprint0) = setup_keys(TPRV0_STR);
+        let locktime_threshold = 500000000; // if less than this means block number, else block time in seconds
+        let locktime_blocks = 100;
+        let locktime_seconds = locktime_blocks + locktime_threshold;
+        let desc = descriptor!(sh (and_v (+v pk prvkey0), (and_v (+v after locktime_seconds), (after locktime_blocks)))).unwrap();
+        let (wallet_desc, keymap) = desc.to_wallet_descriptor(Network::Testnet).unwrap();
+        let signers_container = Arc::new(SignersContainer::from(keymap));
+        let policy = wallet_desc
+            .extract_policy(signers_container)
+            .unwrap()
+            .unwrap();
+
+        println!("desc policy = {:?}", policy); // TODO remove
+
+        // TODO how should this fail with mixed timelocks?
```

Comment:
@afilini when you have a chance I need a little guidance on what you have in mind for this and the following test, I think I have it setup OK but I'm not sure what the expected results should be.

Reply:
That's probably a bug, it should fail when it tries to merge the two timelocks..

As a disclaimer I should say that the whole policy part has received very little testing and I'm definitely expecting to find more bugs going as we add more tests. I also remember seeing weird results when trying with descriptors that represent "time-decaying timelocks", like: `wsh(thresh(pk(A), pk(B), pk(C), older(T_1), older(T_2)))`.

Reply:
thanks for the tip, I'll explore this area more with the thresh descriptors you suggested and see if I can fix it to fail property.

Reply:
Sure, feel free to explore and have a go at it. If you feel like it starts taking too much time, we can also merge this PR as it is now, and then make a new one just for the policy part.

Reply:
I just rebased and this ready to go with descriptor tests for dsl, template, and checksum. Will finish policy testing in a future PR.